mitiq/executor/executor.py

```diff
@@ -137,10 +139,18 @@ def evaluate(
             force_run_all: If True, force every circuit in the input sequence
                 to be executed (if some are identical). Else, detects identical
                 circuits and runs a minimal set.
+
+        Returns:
+            List of real valued expectation values.
         """
         if not isinstance(circuits, List):
             circuits = [circuits]
 
+        if observable is not None and not ishermitian(observable.matrix()):
+            warnings.warn(
+                "Expected observable to be hermitian. Continue with caution."
+            )
```

Comment:
Converting to a matrix is fine only for a small number of qubits. It can become slow for many qubits.
If I am not wrong, a sum of Pauli operators, is Hermitian if and only if all the coefficients are real. 
So we can just check this condition instead of converting to a matrix.

See e.g.: https://github.com/quantumlib/Cirq/blob/506ef08e7368076254ecc4717a804f4a842f441e/cirq-core/cirq/ops/linear_combinations.py#L713

Moreover, what about if we apply `np.real_if_close()` to the final value returned by this function? This would still allow for overloading the function with non-Hermitian observables, but would produce a cleaner output in standard settings.

Reply:
Good catch, and good idea using the coefficients. I've implemented that in [a12e83e](https://github.com/unitaryfund/mitiq/pull/1486/commits/a12e83e7b84366a4a1fdef2629e790ca0c4494a7).

I don't feel like running the results through `np.real_if_close()` is necessary since if a user needs to have complex expectation values, they can still take the real part after using `evaluate`. I don't feel particularly strong about this, however. Happy to do it if you think it's best.

Reply:
> I don't feel like running the results through `np.real_if_close()` is necessary since if a user needs to have complex expectation values, they can still take the real part after using `evaluate`. I don't feel particularly strong about this, however. Happy to do it if you think it's best.

Mmm I am not sure about this. The common way in which the executor will be used is to get the expectation values (of hermitian observables) to be sent through error mitigation techniques (e.g. ZNE) that expects real inputs and produce a mitigated real output. E.g. an exponential fit of data points like (`1.7 + 0.0 j`, `1.6 + 0.0 j `) will fail. 
Basically a complex result is an exception to the rule that we still allow for (with a warning),  but the typical output should probably be float (consistently with the function annotation).

Reply:
You're right, I was thinking the user calls `evaluate` (and hence could be in charge of casting the values to floats), but that's not usually what happens. Thanks for catching that. I've added the `real_if_close` call in [4781a15](https://github.com/unitaryfund/mitiq/pull/1486/commits/4781a155b64fb499ab5137d44759a2863fa386bd).